src/projectvariables.rs

```diff
@@ -0,0 +1,835 @@
+use liquid_core::Value;
+use regex::Regex;
+use thiserror::Error;
+
+use anyhow::Result;
+use liquid::{model::object::Entry, Object};
+
+use crate::config::{Config, TemplateSlotsTable};
+
+#[derive(Debug)]
+pub(crate) struct TemplateSlots {
+    pub(crate) var_name: String,
+    pub(crate) var_info: VarInfo,
+    pub(crate) prompt: String,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum VarInfo {
+    Bool { default: Option<bool> },
+    String { entry: Box<StringEntry> },
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct StringEntry {
+    pub(crate) default: Option<String>,
+    pub(crate) choices: Option<Vec<String>>,
+    pub(crate) regex: Option<Regex>,
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum ConversionError {
+    #[error("parameter `{parameter}` of placeholder `{var_name}` should be a `{correct_type}`")]
+    WrongTypeParameter {
+        var_name: String,
+        parameter: String,
+        correct_type: String,
+    },
+    #[error("placeholder `{var_name}` should be a table")]
+    InvalidPlaceholderFormat { var_name: String },
+    #[error("missing prompt question for `{var_name}`")]
+    MissingPrompt { var_name: String },
+    #[error("choices array empty for `{var_name}`")]
+    EmptyChoices { var_name: String },
+    #[error("default is `{default}`, but is not a valid value in choices array `{choices:?}` for  `{var_name}`")]
+    InvalidDefault {
+        var_name: String,
+        default: String,
+        choices: Vec<String>,
+    },
+    #[error(
+        "invalid type for variable `{var_name}`: `{value}` possible values are `bool` and `string`"
+    )]
+    InvalidVariableType { var_name: String, value: String },
+    #[error("bool type does not support `choices` field")]
+    ChoicesOnBool { var_name: String },
+    #[error("bool type does not support `regex` field")]
+    RegexOnBool { var_name: String },
+    #[error("variable `{var_name}` was missing in config file running on silent mode")]
+    MissingPlaceholderVariable { var_name: String },
+    #[error("field `{field}` of variable `{var_name}` does not match configured regex")]
+    RegexDoesntMatchField { var_name: String, field: String },
+    #[error("regex of `{var_name}` is not a valid regex")]
+    InvalidRegex { var_name: String, regex: String },
+    #[error("placeholder `{var_name}` is not valid as you can't override `project-name` and `crate_name`")]
+    InvalidPlaceholderName { var_name: String },
+}
+
+#[derive(Debug, Clone, PartialEq)]
+enum SupportedVarValue {
+    Bool(bool),
+    String(String),
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum SupportedVarType {
+    Bool,
+    String,
+}
+
+const INVALID_NAMES: [&str; 2] = ["project-name", "crate_name"];
```

Comment:
Maybe the other standard variables should be named here too? Like: `os-arch` and `authors`

Reply:
makes sense to me